plugins/scripting/scripting-compiler/tests/org/jetbrains/kotlin/scripting/compiler/test/ConstructAnnotationTest.kt

```diff
@@ -0,0 +1,132 @@
+import org.jetbrains.kotlin.scripting.compiler.plugin.TestDisposable
+import org.jetbrains.kotlin.scripting.compiler.plugin.impl.ScriptDiagnosticsMessageCollector
+import org.jetbrains.kotlin.scripting.compiler.plugin.impl.createCompilationContextFromEnvironment
+import org.jetbrains.kotlin.scripting.compiler.plugin.impl.getScriptKtFile
+import org.jetbrains.kotlin.scripting.resolve.InvalidScriptResolverAnnotation
+import org.jetbrains.kotlin.scripting.resolve.getScriptCollectedData
+import org.jetbrains.kotlin.test.ConfigurationKind
+import org.jetbrains.kotlin.test.KotlinTestUtils
+import org.jetbrains.kotlin.test.TestJdkKind
+import java.io.File
+import kotlin.reflect.KClass
+import kotlin.script.experimental.api.*
+import kotlin.script.experimental.host.toScriptSource
+import kotlin.script.experimental.jvm.jvm
+
+private const val testDataPath = "plugins/scripting/scripting-compiler/testData/compiler/constructAnnotations"
+
+@Target(AnnotationTarget.FILE)
+@Repeatable
+@Retention(AnnotationRetention.SOURCE)
+private annotation class TestAnnotation(vararg val options: String)
+
+class ConstructAnnotationTest : TestCase() {
+    private val testRootDisposable: Disposable = TestDisposable()
+
+    fun testAnnotationEmptyVarArg() {
+        val annotations = annotations("TestAnnotationEmptyVarArg.kts", TestAnnotation::class)
+            .valueOrThrow()
+            .filterIsInstance(TestAnnotation::class.java)
+
+        assertEquals(annotations.count(), 1)
+        assert(annotations.first().options.isEmpty())
+    }
+
+    fun testBasicVarArgTestAnnotation() {
+        val annotations = annotations("SimpleTestAnnotation.kts", TestAnnotation::class)
+            .valueOrThrow()
+            .filterIsInstance(TestAnnotation::class.java)
+
+        assertEquals(annotations.count(), 1)
+        assertEquals(annotations.first().options.toList(), listOf("option"))
+    }
+
+    fun testAnnotationWithArrayLiteral() {
+        val annotations = annotations("TestAnnotationWithArrayLiteral.kts", TestAnnotation::class)
+            .valueOrThrow()
+            .filterIsInstance(TestAnnotation::class.java)
+
+        assertEquals(annotations.count(), 1)
+        assertEquals(annotations.first().options.toList(), listOf("option"))
+    }
+
+    fun testAnnotationWithArrayOfFunction() {
+        val annotations = annotations("TestAnnotationWithArrayOfFunction.kts", TestAnnotation::class)
+            .valueOrThrow()
+            .filterIsInstance(TestAnnotation::class.java)
+
+        assertEquals(annotations.count(), 1)
+        assertEquals(annotations.first().options.toList(), listOf("option"))
+    }
+
+    fun testAnnotationWithEmptyArrayFunction() {
+        val annotations = annotations("TestAnnotationWithEmptyArrayFunction.kts", TestAnnotation::class)
+            .valueOrThrow()
+            .filterIsInstance(TestAnnotation::class.java)
+
+        assertEquals(annotations.count(), 1)
+        assert(annotations.first().options.isEmpty())
+    }
+
```

Comment:
The missing case is our current `@DependsOn` annotation with a parameter after a `vararg`. It doesn't work with the current argument mapping logic, unfortunately. Let's discuss what to do about it offline.

Reply:
I added a test that successfully constructs this annotation:

```kotlin
@file:AnnotationWithVarArgAndArray("option", moreOptions = ["otherOption"])
```

It will fail if the `moreOptions` label is missing, however, the regular compiler also won't accept it without the label:

From [docs](https://kotlinlang.org/docs/reference/functions.html):
```
Only one parameter may be marked as vararg. If a vararg parameter is not the last one in the list, values for the following parameters can be passed using the named argument syntax, or, if the parameter has a function type, by passing a lambda outside parentheses.
```

I'm not sure if this properly covers the scenario with a lambda outside the parenthesis, but we can't have this in our annotations anyway